src/automatic_dantzig_wolfe.jl

```diff
@@ -0,0 +1,380 @@
+    end
+    o.constraints_to_axes[c] = index_sets
+    o.constraints_to_variables[c] = _get_variables_in_constraint(model, c)
+end
+
+# Computes the index sets of the constraint reference
+function _get_constraint_axes(constraint_ref::AbstractArray)
+    axs = axes(constraint_ref)
+    ds = JC.DenseAxisArray(constraint_ref, axs...)
+    return _get_constraint_axes(ds)
+end
+
+function _get_constraint_axes(constraint_ref::JC.DenseAxisArray)
+    axes_of_constraint = Array{BlockDecomposition.Axis,1}()
+    for a in constraint_ref.axes
+        if a != 1   # Axes of the form 1:1 do not matter (single constraints)
+            push!(axes_of_constraint, Axis(a))
+        end
+    end
+    return axes_of_constraint
+end
+
+function _get_constraint_axes(constraint_ref::JC.SparseAxisArray)
+    indices = eachindex(constraint_ref)
+    axes = Array{Set{Any}}(undef, 1)
+    for index in indices
+        for j in 1:length(index)
+            if length(axes) < length(index)
+                resize!(axes, length(index))
+            end
+            if !isassigned(axes, j)
+                axes[j] = Set()
+            end
+            push!(axes[j], index[j])
+        end
+    end
+    result = Array{BlockDecomposition.Axis,1}()
+    for a in axes
+        push!(result, Axis(a))
+    end
+    return result
+end
+
+function _get_variables_in_constraint(model::JuMP.Model, constraint::JuMP.ConstraintRef)
+    f = MOI.get(model, MathOptInterface.ConstraintFunction(), constraint)
+    variables = Set{MOI.VariableIndex}()
+    for term in f.terms
+        push!(variables, term.variable_index)
+    end
+    return variables
+end
+
+# Computes for the given axes two decomposition structures:
+# In the first one (bs0) all constraints *not indexed by any* axis from
+# axes are in the master, in the second one (bs1), all constraints *indexed
+# by at least one* axis from axes are in the master
+function get_block_structure(
+    axes::Array{<:Axis,1},
+    model_description::ModelDescription,
+    model::JuMP.Model,
+)
+    block_constraints = Set{JuMP.ConstraintRef}()
+    master_constraints = Set{JuMP.ConstraintRef}()
+    for c in keys(model_description.constraints_to_axes)
+        if  isempty(intersect(axes, model_description.constraints_to_axes[c]))
+            push!(master_constraints, c)
+        else
+            push!(block_constraints, c)
+        end
+    end
+    # Create first block structure
+    graph = _create_graph(block_constraints, model_description)
+    blocks = _get_connected_components!(graph)
+    bs0 = BlockStructure(
+        model_description, master_constraints, axes, false, blocks, graph
+    )
+    # Create second block structure (the roles of master constraints and block
+    # constraints are switched)
+    graph = _create_graph(master_constraints, model_description)
+    blocks = _get_connected_components!(graph)
+    bs1 = BlockStructure(
+        model_description, block_constraints, axes, true, blocks, graph
+    )
```

Comment:
Can you call `_get_connected_components!` in the constructor of `BlockStructure` ?

Reply:
Good idea! I changed the code accordingly.